File: python/1376. Time Needed to Inform All Employees.py
class Solution(object):
    def numOfMinutes(self, n, headID, manager, informTime):
        """
        :type n: int
        :type headID: int
        :type manager: List[int]
        :type informTime: List[int]
        :rtype: int
        """
        que = []
        edges = [[] for _ in range(n)]
        for i, head in enumerate(manager):
            if head==-1:
                que.append((i, informTime[i]))
                continue
            edges[head].append(i)
        ans = 0
        while que:
            i, t = que.pop(0)
            ans = max(ans, t)
            for sub in edges[i]:
                que.append((sub, informTime[sub] + t))
        return ans

File: go/src/leetcode/1376/1376. Time Needed to Inform All Employees.go
package main

import "fmt"

func numOfMinutes(n int, headID int, manager []int, informTime []int) (tot int) {
	g := make([][]int, n)
	for i, pi := range manager {
		if pi != -1 {
			g[pi] = append(g[pi], i)
		}
	}
	type pair struct {
		i, t int
	}
	que := []pair{{headID, informTime[headID]}}
	for len(que) != 0 {
		p := que[0]
		que = que[1:]
		tot = max(tot, p.t)
		for _, ni := range g[p.i] {
			que = append(que, pair{ni, p.t + informTime[ni]})
		}
	}

	fmt.Println(tot)
	return
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	numOfMinutes(1, 0, []int{-1}, []int{0})
	numOfMinutes(6, 2, []int{2, 2, -1, 2, 2, 2}, []int{0, 0, 1, 0, 0, 0})
}
